cfpb/cfgov-refresh
# Add schema blocks to answer page

Adds blocks to answer page streamfield to support pilot test of Google FAQ and HowTo schema.

## Additions
- FAQ and HowTo schema blocks and templates

## Changes
- Wrap answer page content blocks in `.row` container divs
- Apply spacing between `.row` divs

## Testing

1. Check out branch
2. Run `gulp styles` and `python cfgov/manage.py migrate`
3. Test FAQ schema:
    - open [an answer page](http://localhost:8000/admin/pages/5764/edit/) in Wagtail admin

    - Add a FAQ schema block in the answer content streamfield:
      <img width="1198" alt="Screen Shot 2019-05-09 at 6 09 35 PM" src="https://user-images.githubusercontent.com/778171/57496121-301e7f80-7286-11e9-9ab2-fa00848eb315.png">

    - Verify that the output is wrapped in schema markup
      <img width="725" alt="Screen Shot 2019-05-09 at 6 12 07 PM" src="https://user-images.githubusercontent.com/778171/57496170-68be5900-7286-11e9-91d0-72c0b64d84cf.png">

    - You can also verify the schema by pasting the FAQ HTML into the [Google structured data testing tool](https://search.google.com/structured-data/testing-tool)
      <img width="687" alt="Screen Shot 2019-05-09 at 6 11 35 PM" src="https://user-images.githubusercontent.com/778171/57496197-88ee1800-7286-11e9-90c7-47e15aa42c29.png">

4. Test HowTo schema
    - open [an answer page](http://localhost:8000/admin/pages/10386/edit/) in Wagtail admin

    - Add a How To schema block in the answer content streamfield:
      <img width="1109" alt="Screen Shot 2019-05-09 at 6 52 05 PM" src="https://user-images.githubusercontent.com/778171/57497235-b8535380-728b-11e9-8584-9b2e6bd7afc6.png">

    - Verify that the output is wrapped in schema markup
      <img width="713" alt="Screen Shot 2019-05-09 at 6 52 46 PM" src="https://user-images.githubusercontent.com/778171/57497236-b8535380-728b-11e9-9f97-f92c01d0a29e.png">
    
    - The HowTo schema also requires some page-level markup that is applied with a feature flag. To verify the complete markup structure in the [Google structured data testing tool](https://search.google.com/structured-data/testing-tool), add a new `HOW_TO_SCHEMA` flag condition for the path of the page you've edited (ie, `^/ask-cfpb/what-is-a-529-savings-plan-en-2074/`). 
      
      <img width="779" alt="Screen Shot 2019-05-10 at 4 42 46 PM" src="https://user-images.githubusercontent.com/778171/57561799-cf08b180-7342-11e9-9366-9d4a019b60df.png">

      Load the answer page, copy its source, and paste it in the tool. You should see something like this:

      <img width="695" alt="Screen Shot 2019-05-10 at 3 50 35 PM" src="https://user-images.githubusercontent.com/778171/57560621-27888080-733c-11e9-89b1-d44333f01bd0.png">

      The top-level `name` value is the page's question.


## Notes
I initially thought about creating a combined `schema` block with a type selector, but the markup for FAQ and HowTo diverged enough in the latest recommendations from Google that a single template for both would be complicated. I think keeping them separate might be more future-proof (and a bit more comprehensible for users, since we can use schema-specific labels like `Steps` vs `Questions`), but I'm also open to looking at the single block approach again if anyone thinks that would be preferable.
